runtime/commonMain/src/kotlinx/serialization/json/JsonParametricSerializer.kt

```diff
@@ -0,0 +1,96 @@
+import kotlinx.serialization.internal.*
+import kotlinx.serialization.modules.*
+import kotlin.reflect.*
+
+/**
+ * Base class for custom serializers that allows selecting polymorphic serializer
+ * without a dedicated class discriminator, on a content basis.
+ *
+ * Usually, polymorphic serialization (represented by [PolymorphicSerializer] and [SealedClassSerializer])
+ * requires a dedicated `"type"` property in JSON stream to
+ * determine actual serializer, which can be used to deserialize Kotlin class. However, sometimes
+ * (often when interacting with external API) such property is not present in the input
+ * and one should guess type using other evidence, e.g. presence of some keys in the object.
+ * This class offers a skeleton implementation of such guessing strategy. Please note that
+ * since JSON content is represented by [JsonElement] class and could be read only with [JsonInput] decoder,
+ * this class works only with [Json] format.
+ *
+ * Deserialization happens in two stages: first, a value from the JSON stream is read
+ * to a [JsonElement]. Then, [selectSerializer] function is called to determine which serializer should be used.
+ * The returned serializer is used to deserialize [JsonElement] back to Kotlin object.
+ *
+ * One could also serialize a value with this serializer. In that case, class discriminator property would not
+ * be added to JSON stream, i.e., deserializing a class from the string and serializing it back yields the original string
+ * However, to determine a serializer, a standard polymorphic mechanism represented by [SerialModule] is used.
+ * For convenience, [serialize] method can lookup default serializer, but it is recommended to follow
+ * standard procedure with [registering][SerializersModuleBuilder.polymorphic].
+ *
+ * Usage example:
+ * ```
+ * interface Payment {
+ *     val amount: String
+ * }
+ *
+ * @Serializable
+ * data class SuccessfulPayment(override val amount: String, val date: String) : Payment
+ *
+ * @Serializable
+ * data class RefundedPayment(override val amount: String, val date: String, val reason: String) : Payment
+ *
+ * object PaymentSerializer : JsonParametricSerializer<Payment>(Payment::class) {
+ *     override fun selectSerializer(element: JsonElement): KSerializer<out Payment> = when {
+ *         "reason" in element -> RefundedPayment.serializer()
+ *         else -> SuccessfulPayment.serializer()
+ *     }
+ * }
+ *
+ * // Now both statements will yield different subclasses of Payment:
+ *
+ * Json.parse(PaymentSerializer, """{"amount":"1.0","date":"03.02.2020"}""")
+ * Json.parse(PaymentSerializer, """{"amount":"2.0","date":"03.02.2020","reason":"complaint"}""")
+ * ```
+ *
+ * @param T A root class for all classes that could be possibly encountered during serialization and deserialization.
+ * @param baseClass A class token for [T].
+ *
+ */
+public abstract class JsonParametricSerializer<T : Any>(private val baseClass: KClass<T>) : KSerializer<T> {
```

Comment:
Maybe let's rename it to `JsonContextualSerializer`?
Not sure how approachable is `parametric`

Reply:
Well, I though 'Parametric Polymorphsim' is somewhat used term in Jackson and Gson but I've looked it up and seems it is not used anywhere outside FP. Anyway, 'parametric' kinda may fit because we look at unknown in advance parameter – json property

Contextual doesn't fit either because we have regular `@Contextual` and `ContextSerializer` that work in a completely different way to `JsonContextualSerializer`

All references say something about 'property value' or 'property presence' or 'property definitions'... Maybe we should coin something around words 'property' , 'content' or 'value'

Reply:
Let's stick with parametric then